### PR TITLE
[DDS-1058] : PHP-FPM worker pool optimisations

### PR DESCRIPTION
**Pull request by FleetAdmiralButter**
# Checklist
- [x] Affected Issues have been mentioned in the Closing issues section
- [x] Documentation has been written/updated.
- [x] Changelog entry has been written

Explain the **details** for making this change. What existing problem does the pull request solve?

# Changelog Entry

Improvement - Configures PHP-FPM to use a static pool of 24 workers. This improves performance when experiencing traffic spikes by reducing the CPU usage used for forking new processes. New requests are also served more quickly since they are not blocked waiting for new PHP processes to start. (DDS-1058)

### CODE REVIEW

**Comment by FleetAdmiralButter** [comment]
== Gradual Ramp-Up ==

Performance using pm = static, 50 workers, HPA up to 5 pods:

```
running (7m03.9s), 000/100 VUs, 9077 complete and 0 interrupted iterations
default ✓ [======================================] 000/100 VUs  7m0s

     ✗ GET was 200
      ↳  99% — ✓ 36307 / ✗ 1
     ✓ JSONAPI returned data
     ✓ No error returned by JSONAPI

     checks.........................: 99.99% ✓ 108921    ✗ 1
     data_received..................: 2.2 GB 5.2 MB/s
     data_sent......................: 16 MB  37 kB/s
     http_req_blocked...............: avg=3.72ms   min=300ns    med=500ns    max=1.12s    p(90)=600ns    p(95)=48.3ms
     http_req_connecting............: avg=1.8ms    min=0s       med=0s       max=1.1s     p(90)=0s       p(95)=21.4ms
   ✗ http_req_duration..............: avg=280.83ms min=118.6ms  med=263.91ms max=4.06s    p(90)=363.43ms p(95)=408.39ms
       { expected_response:true }...: avg=280.83ms min=151.4ms  med=263.91ms max=4.06s    p(90)=363.43ms p(95)=408.4ms
   ✓ http_req_failed................: 0.00%  ✓ 1         ✗ 36307
     http_req_receiving.............: avg=32.41ms  min=0s       med=25.02ms  max=300.4ms  p(90)=68.5ms   p(95)=72.08ms
     http_req_sending...............: avg=175.66µs min=15.2µs   med=22.5µs   max=121.2ms  p(90)=39.42µs  p(95)=66.8µs
     http_req_tls_handshaking.......: avg=1.77ms   min=0s       med=0s       max=466.41ms p(90)=0s       p(95)=25.74ms
     http_req_waiting...............: avg=248.24ms min=118.58ms med=227.01ms max=4.01s    p(90)=313.22ms p(95)=357.44ms
     http_reqs......................: 36308  85.654349/s
     iteration_duration.............: avg=3.14s    min=810.63ms med=3.08s    max=8.15s    p(90)=4.15s    p(95)=4.27s
     iterations.....................: 9077   21.413587/s
     vus............................: 5      min=1       max=100
     vus_max........................: 100    min=100     max=100
```

Memory use peaked at about ~700MB per pod:
image

**Comment by FleetAdmiralButter** [comment]
== Gradual Ramp-Up ==

Performance using pm = dynamic, 2-50 workers, HPA up to 4 pods:

```
running (7m03.6s), 000/100 VUs, 8690 complete and 0 interrupted iterations
default ✓ [======================================] 000/100 VUs  7m0s

     ✓ GET was 200
     ✓ JSONAPI returned data
     ✓ No error returned by JSONAPI

     checks.........................: 100.00% ✓ 104280    ✗ 0
     data_received..................: 2.1 GB  5.0 MB/s
     data_sent......................: 11 MB   26 kB/s
     http_req_blocked...............: avg=951.92µs min=300ns    med=500ns    max=1.12s    p(90)=500ns    p(95)=600ns
     http_req_connecting............: avg=416.59µs min=0s       med=0s       max=1.09s    p(90)=0s       p(95)=0s
   ✗ http_req_duration..............: avg=321.83ms min=151.81ms med=270.51ms max=8.85s    p(90)=432.22ms p(95)=594.79ms
       { expected_response:true }...: avg=321.83ms min=151.81ms med=270.51ms max=8.85s    p(90)=432.22ms p(95)=594.79ms
   ✓ http_req_failed................: 0.00%   ✓ 0         ✗ 34760
     http_req_receiving.............: avg=40.41ms  min=29.5µs   med=22.42ms  max=3.67s    p(90)=70.46ms  p(95)=82.17ms
     http_req_sending...............: avg=112.54µs min=14.9µs   med=22.4µs   max=297.02ms p(90)=32.3µs   p(95)=39.6µs
     http_req_tls_handshaking.......: avg=473.27µs min=0s       med=0s       max=514.19ms p(90)=0s       p(95)=0s
     http_req_waiting...............: avg=281.3ms  min=138.01ms med=239.74ms max=8.77s    p(90)=373.97ms p(95)=500.61ms
     http_reqs......................: 34760   82.065618/s
     iteration_duration.............: avg=3.28s    min=1.83s    med=3.13s    max=16.86s   p(90)=4.32s    p(95)=4.73s
     iterations.....................: 8690    20.516405/s
     vus............................: 6       min=1       max=100
     vus_max........................: 100     min=100     max=100
```

Memory use peaked ~300MB:
image

**Comment by FleetAdmiralButter** [comment]
==Load Spike Scenario==
0-100VUs in 10 seconds, 100VUs for 1m, 100-0VUs in 10s
Performance using pm = static, 50 workers, HPA up to 5 pods:

```
running (1m23.8s), 000/100 VUs, 2034 complete and 0 interrupted iterations
default ✓ [======================================] 000/100 VUs  1m20s

     ✓ GET was 200
     ✓ JSONAPI returned data
     ✓ No error returned by JSONAPI

     checks.........................: 100.00% ✓ 24408     ✗ 0
     data_received..................: 488 MB  5.8 MB/s
     data_sent......................: 2.5 MB  30 kB/s
     http_req_blocked...............: avg=739.3µs  min=300ns   med=500ns    max=1.07s    p(90)=500ns    p(95)=600ns
     http_req_connecting............: avg=412.53µs min=0s      med=0s       max=1.04s    p(90)=0s       p(95)=0s
   ✗ http_req_duration..............: avg=381.45ms min=149.6ms med=311.48ms max=5.45s    p(90)=546.21ms p(95)=770.88ms
       { expected_response:true }...: avg=381.45ms min=149.6ms med=311.48ms max=5.45s    p(90)=546.21ms p(95)=770.88ms
   ✓ http_req_failed................: 0.00%   ✓ 0         ✗ 8136
     http_req_receiving.............: avg=34.72ms  min=30.8µs  med=26.24ms  max=519.38ms p(90)=69.53ms  p(95)=74.78ms
     http_req_sending...............: avg=24.81µs  min=15.5µs  med=22.1µs   max=316.9µs  p(90)=32.25µs  p(95)=39.19µs
     http_req_tls_handshaking.......: avg=323.65µs min=0s      med=0s       max=44.12ms  p(90)=0s       p(95)=0s
     http_req_waiting...............: avg=346.7ms  min=137.9ms med=275.78ms max=5.38s    p(90)=493.49ms p(95)=719.8ms
     http_reqs......................: 8136    97.102426/s
     iteration_duration.............: avg=3.52s    min=1.88s   med=3.4s     max=9.09s    p(90)=4.68s    p(95)=5.28s
     iterations.....................: 2034    24.275607/s
     vus............................: 2       min=2       max=100
     vus_max........................: 100     min=100     max=100
```

**Comment by FleetAdmiralButter** [comment]
==Load Spike Scenario==
0-100VUs in 10 seconds, 100VUs for 1m, 100-0VUs in 10s
Performance using pm = dynamic, 2-50 workers, HPA up to 5 pods:

```
running (1m22.4s), 000/100 VUs, 2124 complete and 0 interrupted iterations
default ✓ [======================================] 000/100 VUs  1m20s

     ✓ GET was 200
     ✓ JSONAPI returned data
     ✓ No error returned by JSONAPI

     checks.........................: 100.00% ✓ 25488      ✗ 0
     data_received..................: 515 MB  6.2 MB/s
     data_sent......................: 3.4 MB  42 kB/s
     http_req_blocked...............: avg=3.36ms   min=300ns    med=500ns    max=1.11s    p(90)=500ns    p(95)=44.34ms
     http_req_connecting............: avg=1.66ms   min=0s       med=0s       max=1.08s    p(90)=0s       p(95)=0s
   ✓ http_req_duration..............: avg=340.3ms  min=162.2ms  med=297.43ms max=3.95s    p(90)=500.7ms  p(95)=640.99ms
       { expected_response:true }...: avg=340.3ms  min=162.2ms  med=297.43ms max=3.95s    p(90)=500.7ms  p(95)=640.99ms
   ✓ http_req_failed................: 0.00%   ✓ 0          ✗ 8496
     http_req_receiving.............: avg=29.25ms  min=30.3µs   med=23.45ms  max=309.14ms p(90)=70.04ms  p(95)=74.3ms
     http_req_sending...............: avg=107.27µs min=15.2µs   med=22.3µs   max=125.37ms p(90)=37.29µs  p(95)=57.82µs
     http_req_tls_handshaking.......: avg=1.55ms   min=0s       med=0s       max=169.6ms  p(90)=0s       p(95)=0s
     http_req_waiting...............: avg=310.94ms min=143.06ms med=263.6ms  max=3.92s    p(90)=462.34ms p(95)=601.58ms
     http_reqs......................: 8496    103.048561/s
     iteration_duration.............: avg=3.38s    min=1.8s     med=3.28s    max=7.8s     p(90)=4.46s    p(95)=4.73s
     iterations.....................: 2124    25.76214/s
     vus............................: 1       min=1        max=100
     vus_max........................: 100     min=100      max=100
```

**Comment by FleetAdmiralButter** [comment]
Reduced to 24 workers:

image

370MB per pod is more in line with how current containers are behaving

Current containers will consume between 140MB and 700MB depending on traffic:

Idle:
image

Load + actively scaling:
image

**Comment by FleetAdmiralButter** [comment]
Change documented at https://digital-engagement.atlassian.net/wiki/spaces/DDS/pages/3023405108/FPM+worker+configuration.

File: bay/images/Dockerfile.php [hunk 1]
```diff
@@ -14,6 +14,10 @@
 COPY docker/redis-unavailable.services.yml /bay
 COPY docker/settings.php /bay
 
+# Configure a sane worker pool
```
Comment (nicksantamaria) [inline bay/images/Dockerfile.php]:
```suggestion
# Change worker pool from dynamic to static. Change default value to 24.
```

**Review by nicksantamaria — APPROVED** [review]

**Review by nicksantamaria — APPROVED** [review]

**Comment by FleetAdmiralButter** [comment]
Tested locally, ClamAV service still works as expected.

**Pull request merged**